torch/csrc/jit/tensorexpr/bounds_overlap.cpp

```diff
@@ -6,6 +6,73 @@ namespace jit {
 namespace tensorexpr {
 namespace analysis {
 
+// Returns true if the given expression includes a "Sub" term.
+bool hasSub(ExprPtr e) {
+  return !NodeFinder<Sub>::find(e).empty();
+}
+
+// Returns true if the given expression includes a "Div" term.
+bool hasDiv(ExprPtr e) {
+  return !NodeFinder<Div>::find(e).empty();
+}
+
+// Returns true if all immediates in the given expression are positive.
+bool allImmsArePositive(ExprPtr e) {
+#define TYPE_CASE(Type, Name)                                 \
+  {                                                           \
+    auto imms = NodeFinder<Name##Imm>::find(e);               \
+    for (const auto& imm : imms) {                            \
+      if (immediateIsNegative(imm) || immediateIsZero(imm)) { \
+        return false;                                         \
+      }                                                       \
+    }                                                         \
+  }
+  AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TYPE_CASE);
+#undef TYPE_CASE
+  {
+    auto imms = NodeFinder<BoolImm>::find(e);
```

Comment:
Just curious, why do we care about `BoolImm`s in arithmetic expressions?

Reply:
Only because a bool being 0 can make the overall expression 0.